jnigen/test/simple_package_test/runtime_test_registrant.dart

```diff
@@ -528,6 +529,59 @@ void registerTests(String groupName, TestRunnerCallback test) {
       });
     });
   });
+
+  group('interface implementation', () {
+    test('MyInterface.implement consume on another thread', () async {
+      final voidCallbackResult = Completer<JString>();
+      final myInterface = MyInterface.implement(
+        voidCallback: (s) {
+          voidCallbackResult.complete(s);
+        },
+        stringCallback: (s) {
+          return (s.toDartString(deleteOriginal: true) * 2).toJString();
+        },
+        varCallback: (t) {
+          return "".toJString();
+        },
+        manyPrimitives: (a, b, c, d) {
+          return 0;
+        },
+        T: JString.type,
+      );
+      MyInterfaceConsumer.consumeOnAnotherThread(
+        myInterface,
+        'hello'.toJString(),
+      );
+      final result = await voidCallbackResult.future;
+      expect(result.toDartString(deleteOriginal: true), 'hellohello');
+      myInterface.delete();
+    });
+    test('MyInterface.implement consume on the same thread', () async {
+      final voidCallbackResult = Completer<JString>();
+      final myInterface = MyInterface.implement(
```

Comment:
I believe we should try to make it an actual interface implementation rather than passing closures for everything.

```dart
abstract class MyInterface {
  void Function voidCallback(jni.JString s);
  jni.JString stringCallback(jni.JString s);
  // ..
}
```

And then you'd implement it:

```dart
class MyInterfaceImplementation extends MyInterface {
  final voidCallbackResult = Completer<JString>();

  void Function voidCallback(jni.JString s){
    voidCallbackResult.complete();
  }

  // ...
}
```

The implementation would then be instead of holding on to `_$methods` hold on to an object and use tearoffs rather than closures.

I believe this would be a better API and better performance.
Is there something I've missed about why we do it closure based?